source/common/upstream/upstream_impl.h

```diff
@@ -107,6 +107,8 @@ class HostDescriptionImpl : virtual public HostDescription,
   void metadata(MetadataConstSharedPtr new_metadata) override {
     absl::WriterMutexLock lock(&metadata_mutex_);
     metadata_ = new_metadata;
+    // Update data members dependent on metadata.
```

Comment:
Could  resolveTransportSocketFactory be moved out of the lock guard?
The referenced Metadata should be immutable and valid.

Reply:
@lambdai I can add a scope around the lock to unlock it before calling `resolveTransportSocketFactory`. Why do you think it is necessary? Deadlock?

Reply:
Sorry for the late reply.
It's not necessary but nice to have. 
The only reason is performance. `resolveTransportSocketFactory` doesn't need the lock to protecting

Reply:
Take it or leave it. It's probably not a bottleneck.

auto factory = resolveTransportSocketFactory(address_, new_metadata.get());
{
 absl::WriterMutexLock lock(&amp;metadata_mutex_);    
 metadata_ = new_metadata;
 socket_factory_ = factory;
}